piptools/utils.py

```diff
@@ -591,33 +591,44 @@ def _validate_config(
     :raises click.NoSuchOption: if config contains unknown keys.
     :raises click.BadOptionUsage: if config contains invalid values.
     """
-    cli_params = {
-        param.name: param
-        for param in click_context.command.params
-        if param.name is not None
+    from piptools.scripts.compile import cli as compile_cli
+    from piptools.scripts.sync import cli as sync_cli
+
+    compile_cli_params = {
+        param.name: param for param in compile_cli.params if param.name is not None
+    }
+
+    sync_cli_params = {
+        param.name: param for param in sync_cli.params if param.name is not None
     }
 
+    all_keys = set(compile_cli_params) | set(sync_cli_params)
+
     for key, value in config.items():
-        # Validate unknown keys
-        if key not in cli_params:
-            possibilities = difflib.get_close_matches(key, cli_params.keys())
+        # Validate unknown keys in both compile and sync
+        if key not in all_keys:
+            possibilities = difflib.get_close_matches(key, all_keys)
             raise click.NoSuchOption(
                 option_name=key,
                 message=f"No such config key {key!r}.",
                 possibilities=possibilities,
                 ctx=click_context,
             )
 
-        # Validate invalid values
-        param = cli_params[key]
-        try:
-            param.type.convert(value=value, param=param, ctx=click_context)
-        except Exception as e:
-            raise click.BadOptionUsage(
-                option_name=key,
-                message=f"Invalid value for config key {key!r}: {value!r}.",
-                ctx=click_context,
-            ) from e
+        # Validate values for both compile and sync
+        for cli_params in (compile_cli_params, sync_cli_params):
+            param = cli_params.get(key)
+            if param is None:
+                continue
```

Comment:
Do we need this? I think wrong keys are caught above?

Reply:
yeah, but we still checking both cli_params, see 618 line.

Reply:
Sorry if I understood you wrong, but isn't this covered by the `all_keys` variable test? Also, the value can't possibly be `None`?

Reply:
Suppose there is an `ask = true` option in the config. It's included in `all_keys`, but not found within  `compile_cli_params`. This Is why we skip it and proceed to validate it on `sync_cli_params`.

The condition `if key not in all_keys` is True if there is a unknown key that neither `pip-compile` nor `pip-sync` has.

Reply:
Thank you for the explanantion! I actually thought both dicts were combined into one.

Reply:
Considering your questions, I've refactored a bit in 5b029cd. Does it look all right?